core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -407,30 +419,106 @@ public void index(Index index) {
-                deleted = currentVersion == Versions.NOT_FOUND;
+            final VersionValue versionValue;
+            // if anything is fishy here ie. there is a retry we go and force updateDocument below so we are updating the document in the
+            // lucene index without checking the version map but we still do the version check
+            final boolean forceUpdateDocument;
+            if (canOptimizeAddDocument(index)) {
+                long deOptimizeTimestamp = maxUnsafeAutoIdTimestamp.get();
+                if (index.isRetry()) {
+                    forceUpdateDocument = true;
+                    do {
+                        deOptimizeTimestamp = maxUnsafeAutoIdTimestamp.get();
+                        if (deOptimizeTimestamp >= index.getAutoGeneratedIdTimestamp()) {
+                            break;
+                        }
+                    } while(maxUnsafeAutoIdTimestamp.compareAndSet(deOptimizeTimestamp,
+                        index.getAutoGeneratedIdTimestamp()) == false);
+                    assert maxUnsafeAutoIdTimestamp.get() >= index.getAutoGeneratedIdTimestamp();
+                } else {
+                    // in this case we force
+                    forceUpdateDocument = deOptimizeTimestamp >= index.getAutoGeneratedIdTimestamp();
+                }
+                versionValue = null;
+                currentVersion = Versions.NOT_FOUND;
+                deleted = index.version() == Versions.MATCH_DELETED;
             } else {
-                currentVersion = checkDeletedAndGCed(versionValue);
-                deleted = versionValue.delete();
+                // update the document
+                forceUpdateDocument = false; // we don't force it - it depends on the version
+                versionValue = versionMap.getUnderLock(index.uid());
+                assert incrementVersionLookup();
+                if (versionValue == null) {
+                    currentVersion = loadCurrentVersionFromIndex(index.uid());
+                    deleted = currentVersion == Versions.NOT_FOUND;
+                } else {
+                    currentVersion = checkDeletedAndGCed(versionValue);
+                    deleted = versionValue.delete();
+                }
             }
-
             final long expectedVersion = index.version();
             if (checkVersionConflict(index, currentVersion, expectedVersion, deleted)) {
                 index.setCreated(false);
                 return;
             }
-
             final long updatedVersion = updateVersion(index, currentVersion, expectedVersion);
-
-            indexOrUpdate(index, currentVersion, versionValue);
-
+            if (currentVersion == Versions.NOT_FOUND && forceUpdateDocument == false) {
+                // document does not exists, we can optimize for create
+                index.setCreated(true);
+                index(index, indexWriter);
+            } else {
+                update(index, versionValue, indexWriter);
```

Comment:
there is a problem here with the index.setCreated flag when ends up being false on retry for auto gen id docs. I think we can move `index.setCreated` out of this if and do `index.setCreated(deleted)` just before the if (as a side effect, it will mean we don't need to pass `versionValue` to `update()` and can reduce the scope of that variable.